change and fixes for existing funcs

1e-to opened the pull request:
For now in this PR I did following:
1)	Fix a typo with the name of one of the getter `DPPLDevice_GetMaxComputeUnites` to `DPPLDevice_GetMaxComputeUnits` 
2)	Add the function declarations for the getters to backend.pxd. 
3)	Add dppl_utils.h/DPPLSize_t_Array_Delete to backend.pxd.
4)	Add members and getters for all the attributes to SyclDevice class in sycl_core.pyx and _sycl_core.pxd.
5)	Also change the existing getters get_device_name, get_device_type, get_vendor_name, get_driver_version also to cpdef.
6)	Add unit tests for the Cython functions to a new test_sycl_device.py file. 

WIP

PokhodenkoSA committed:
Use the black badge in project's README.md

See https://github.com/psf/black#show-your-style

Commit:
A current device is always present even outside a device context.

     -- default device is always available.

Commit:
Allow program creation only for OpenCL for the time being.

Commit:
Fix errors found for Sergey.

Commit:
Add support for uint8 and ulong.

Commit:
Add support for uint8 and ulong.

Commit:
Remove useless except+, as C++ exceptions do not cross C API.

oleksandr-pavlyk committed:
no need to use enum in return type

oleksandr-pavlyk committed:
Fixed TBB warnings on setting GPU queue

Solution is not to create context on devices that are not
of the requested device type or not from the requested backend.

```
Python 3.7.7 (default, Jul 14 2020, 22:02:37)
Type 'copyright', 'credits' or 'license' for more information
IPython 7.17.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import dpctl

In [2]: with dpctl.device_context('opencl:gpu'):
   ...:     print(1)
      ...:
      1

In [3]: with dpctl.device_context('level0:gpu'):
   ...:     print(1)
      ...:
      1

```

Commit:
Rename DPPLSyclBEType to DPPLSyclBackendType. Drop superfluous enum specifier.

Commit:
Merge branch 'feature/backend_specific_queues_v2' of github.com:diptorupd/dpctl into feature/backend_specific_queues_v2

Commit:
Reformat.

Commit:
Fix formatting.

PokhodenkoSA committed:
onaAPI beta10 compiler (#103)

* CC=clang.exe

* CC=clang-cl.exe

PokhodenkoSA committed:
onaAPI beta10 compiler (#103)

* CC=clang.exe

* CC=clang-cl.exe

diptorupd committed:
Merge pull request #93 from diptorupd/feature/backend_specific_queues_v2

Feature/backend specific queues v2

Commit:
Merge branch 'fix/issue_83' of github.com:diptorupd/dpctl into fix/issue_83

Commit:
Merge branch 'diptorupd-fix/issue_83' into upstream_master

diptorupd committed:
Fix/issue 83 (#96)

* Add support for uint8 and ulong.

Commit:
Merge branch 'master' of github.com:IntelPython/dpctl into upstream_master

diptorupd committed:
Feature/get ref address (#105)

* Rename all functions to addressof_ref and add docstring.

Co-authored-by: reazul.hoque <reazul.hoque@intel.com>

PokhodenkoSA committed:
Rename PyDPPL to dpCtl in all visible places (#104)

* Rename readme and headers

* Fix left over places where DPPL/PyDPPL was used in comments.

Co-authored-by: Diptorup Deb <diptorup.deb@intel.com>

PokhodenkoSA committed:
Fix bugs highlighted by tools (#92)

* Fix unhandled character

* Fix unhandled character

* Remove unused imports

* Fix class name duplication

PokhodenkoSA committed:
Update README.md

Warning about conda-build=3.20

diptorupd committed:
Change cast from long to size_t (#111)

1e-to committed:
Add changelog (#108)

* Add changelog

Co-authored-by: etotmeni <elena.totmenina@intel.com>
Co-authored-by: Diptorup Deb <3046810+diptorupd@users.noreply.github.com>

diptorupd committed:
Update CONTRIBUTING.md

oleksandr-pavlyk committed:
Use INSTALL_PATH in cmake call

This ensures that backend libraries  are copied into dpctl
and included in the conda tar ball.

Also fixed copy statement to account for .lib files being
in %INSTALL_PREFIX%\lib and .dll files being in
%INSTALL_PREFIX%\bin on Windows.

oleksandr-pavlyk committed:
Merge pull request #114 from IntelPython/fix-windows-script

Use INSTALL_PATH in cmake call

diptorupd committed:
Update CHANGELOG.md

oleksandr-pavlyk committed:
Fixed typo in the code for set_default_queue (#117)

* Fixed typo in the code for set_default_queue

Allowing both strings or enums as selectors for
backend and device.

* Fixed "variabel reference before assignment" in device_context manager

ctxt must be set outside of try/finally to be guaranteed avaialble
in the finally clause.

PokhodenkoSA committed:
Use only first device for each device type and backend (#118)

diptorupd committed:
Fix failing platform test case. (#116)

* Fix failing platform test case.

* Skip if no non-host back end is there.

* Remove extra new line.

* change how the initial active queue is populated.

* It is fine for the queue manager to store the host queue. Up to clients what they want to do with it.

* Fix gtests.

* Add try catch...

* handle exception.

* Skip tests when no OpenCL devices are available.

* Skip more tests if no devices exist.

* Skip Python tests.

* Skip Python test if no usable device exists.

* Only report number of non-host platforms.

* Add skip tests to test_sycl_usm.py

* Skip more tests if no platform is available.

* Properly fix initialization of static variable.

* Do not create new std::vector.

* Raise exception if DPPLSyclQueueRef is NULL.

Commit:
Rename all pxd and pyx filenames to start with _

Commit:
change and fixes for existing funcs

Commit:
Add device::aspects for int64_base_atomics and int64_extended_atomics

oleksandr-pavlyk committed:
dpctl.get_current_queue is now cimportable

diptorupd commented on dpctl/sycl_core.pyx:
```diff
@@ -145,19 +150,62 @@ cdef class SyclDevice:
+        max_work_item_sizes = []
+        for n in range(3):
+            max_work_item_sizes.append(self._max_work_item_sizes[n])
+        DPPLSize_t_Array_Delete(self._max_work_item_sizes)
```
Should the DPPLSize_t_Array_Delete call be here? I think should be moved to the `__dealloc__` otherwise the `self._max_work_item_sizes` is deleted after first call to `get_max_work_item_sizes`.

diptorupd commented on dpctl/tests/test_sycl_device.py:
```diff
@@ -0,0 +1,72 @@
+
+    def test_get_max_compute_units (self):
+        q = dpctl.get_current_queue()
+        try:
```
Also test if max_compute_units is returned some expected value and not garbage?

diptorupd commented on dpctl/tests/test_sycl_device.py:
```diff
@@ -0,0 +1,72 @@
+class TestSyclDevice (unittest.TestCase):
+
+    def test_get_max_compute_units (self):
+        q = dpctl.get_current_queue()
```
Maybe check for exception on `get_current_queue` as well since that call can raise a `SyclQueueCreationError`.

diptorupd commented on dpctl/tests/test_sycl_device.py:
```diff
@@ -0,0 +1,72 @@
+    def test_get_max_work_item_dims (self):
+        q = dpctl.get_current_queue()
+        try:
+            max_work_item_dims = q.get_sycl_device().get_max_work_item_dims()
```
same.

diptorupd commented on dpctl/tests/test_sycl_device.py:
```diff
@@ -0,0 +1,72 @@
+    def test_get_max_work_item_sizes (self):
+        q = dpctl.get_current_queue()
+        try:
+            max_work_item_sizes = q.get_sycl_device().get_max_work_item_sizes()
```
same.

diptorupd commented on dpctl/tests/test_sycl_device.py:
```diff
@@ -0,0 +1,72 @@
+    def test_get_max_work_group_size (self):
+        q = dpctl.get_current_queue()
+        try:
+            max_work_group_size = q.get_sycl_device().get_max_work_group_size()
```
same.

diptorupd commented on dpctl/tests/test_sycl_device.py:
```diff
@@ -0,0 +1,72 @@
+    def test_get_max_num_sub_groups (self):
+        q = dpctl.get_current_queue()
+        try:
+            max_num_sub_groups = q.get_sycl_device().get_max_num_sub_groups()
```
same.

diptorupd commented on backends/include/dppl_sycl_device_interface.h:
```diff
@@ -152,6 +152,26 @@ DPPL_API
+ */
+DPPL_API
+bool
+DPPLDevice_GetAspectsBaseAtomics (__dppl_keep const DPPLSyclDeviceRef DRef);
```
Let us rename to `DPPLDevice_HasInt64BaseAtomics`.

diptorupd commented on backends/include/dppl_sycl_device_interface.h:
```diff
@@ -152,6 +152,26 @@ DPPL_API
+ */
+DPPL_API
+bool
+DPPLDevice_GetAspectsExtendedAtomics (__dppl_keep const DPPLSyclDeviceRef DRef);
```
let us rename to `DPPLDevice_HasInt64ExtendedAtomics`

diptorupd commented on backends/source/dppl_sycl_device_interface.cpp:
```diff
@@ -196,6 +196,26 @@ DPPLDevice_GetMaxNumSubGroups (__dppl_keep const DPPLSyclDeviceRef DRef)
+{
+    auto D = unwrap(DRef);
+    if(D) {
+        return D->has(aspect::int64_base_atomics);
```
I did not know about the functions. Nice!

diptorupd commented on backends/tests/test_sycl_device_interface.cpp:
```diff
@@ -175,6 +175,42 @@ TEST_F (TestDPPLSyclDeviceInterface, CheckOCLGPU_GetMaxNumSubGroups)
+    if(!OpenCL_cpu)
+        GTEST_SKIP_("Skipping as no OpenCL CPU device found.");
+
+    auto n = DPPLDevice_GetAspectsBaseAtomics(OpenCL_cpu);
```
What do we expect on devices that actually do not have `int64_base_atomics`? In such cases `false` is the right result. I think a better way to test this is to include `CL::sycl.hpp` and call the get the aspect for the device using SYCL and see if it matches what DPPLDevice_xxx returns. It is not really a great test, just sanity check.

diptorupd commented on backends/tests/test_sycl_device_interface.cpp:
```diff
@@ -175,6 +175,42 @@ TEST_F (TestDPPLSyclDeviceInterface, CheckOCLGPU_GetMaxNumSubGroups)
+
+    auto n = DPPLDevice_GetAspectsBaseAtomics(OpenCL_gpu);
+    EXPECT_TRUE(n != false);
+}
```
Same here.

diptorupd commented on backends/tests/test_sycl_device_interface.cpp:
```diff
@@ -175,6 +175,42 @@ TEST_F (TestDPPLSyclDeviceInterface, CheckOCLGPU_GetMaxNumSubGroups)
+        GTEST_SKIP_("Skipping as no OpenCL CPU device found.");
+
+    auto n = DPPLDevice_GetAspectsExtendedAtomics(OpenCL_cpu);
+    EXPECT_TRUE(n != false);
```
Same here.

diptorupd commented on backends/tests/test_sycl_device_interface.cpp:
```diff
@@ -175,6 +175,42 @@ TEST_F (TestDPPLSyclDeviceInterface, CheckOCLGPU_GetMaxNumSubGroups)
+        GTEST_SKIP_("Skipping as no OpenCL GPU device found.");
+
+    auto n = DPPLDevice_GetAspectsExtendedAtomics(OpenCL_gpu);
+    EXPECT_TRUE(n != false);
```
and here.

diptorupd commented on dpctl/tests/test_sycl_device.py:
```diff
@@ -68,5 +68,21 @@ def test_get_max_num_sub_groups (self):
 
+    def test_get_aspects_base_atomics (self):
+        q = dpctl.get_current_queue()
+        try:
```
Aha now I see. Basically we are testing that the function should never raise an exception. OK.

Commit:
Fixes according comments

Commit:
Fix tests

PokhodenkoSA committed:
Add GitHub Actions for black

This GitHub Action checks that code in pull request
is formatted with black.

See https://black.readthedocs.io/en/stable/github_actions.html

PokhodenkoSA committed:
Format all Python code with black

Use black version 20.8b1 or branch [stable](https://github.com/psf/black/tree/stable)).

Command: black .

PokhodenkoSA committed:
Ignore versioneer.py from black formatting

versioneer.py is not a part of the project.

Add pyproject.toml file with section [tool.black] and parameter
exclude = 'versioneer.py'. This file will be used by developers and
GitHub Actions for black.

Usign pyproject.toml allows use black like this: `black .`

PokhodenkoSA committed:
Add ignoring revisions for blame

File .git-blame-ignore-revs contains a list of revisions to use for
ignoring in blame. Example:
$ git blame FILE --ignore-revs-file .git-blame-ignore-revs

Also you can configure git to automatically ignore revisions listed
in a file on every call to git blame:
$ git config blame.ignoreRevsFile .git-blame-ignore-revs

Requirements: Git >=2.23

PokhodenkoSA committed:
Update CONTRIBUTING.md

PokhodenkoSA committed:
Merge pull request #91 from PokhodenkoSA/tools/black

Add black formatter for Python code

oleksandr-pavlyk committed:
Merge pull request #125 from IntelPython/expose-get-current-queue-on-C-side

dpctl.get_current_queue is now cimportable

PokhodenkoSA commented on backends/include/dppl_sycl_device_interface.h:
```diff
@@ -110,10 +110,10 @@ DPPLDevice_GetDriverInfo (__dppl_keep const DPPLSyclDeviceRef DRef);
 DPPL_API
 uint32_t
-DPPLDevice_GetMaxComputeUnites (__dppl_keep const DPPLSyclDeviceRef DRef);
+DPPLDevice_GetMaxComputeUnits (__dppl_keep const DPPLSyclDeviceRef DRef);
```
It is very dangerous to rename API function. It is better to create a new one with new name and implement the old one via the new one.

PokhodenkoSA replied:
Then mark it as deprecated. Better to print warning in stderr when old function is used.
And remove in the next release.

Commit:
merge master

Commit:
Codestyle

Commit:
Codestyle

Commit:
Some fixes + add tests for level0 gpu

Commit:
Merge branch 'upstream_master'

Commit:
Change names in tests level0gpu

PokhodenkoSA committed:
Add flags for fixing hardening-check (#149)

See https://wiki.debian.org/HardeningWalkthrough#Testing_your_packages_after_conversion

1e-to committed:
Remove OpenCL interface (#138)

* Del old ocl interface

* Del cffi

* Fix CMakeLists

* Remove OpenCL from README.md

* Update changelog

Co-authored-by: etotmeni <elena.totmenina@intel.com>
Co-authored-by: Sergey Pokhodenko <sergey.pokhodenko@intel.com>

Commit:
Merge branch 'upstream_master'

Commit:
merge master

Commit:
Update changelog

oleksandr-pavlyk committed:
fixed filename typo in setup.py to fix build

PokhodenkoSA committed:
Merge branch 'master' into extra_device_info

PokhodenkoSA committed:
Fix import _backend in _sycl_core.pxd
